app/javascript/lib/validation/validate-file-content.js

```diff
@@ -399,6 +478,25 @@ function validateClusterCoordinates(headers) {
   return issues
 }
 
+
+/** parse a dense matrix file */
+export async function parseDenseMatrixFile(chunker, mimeType, fileOptions) {
+  const { headers, delimiter } = await getParsedHeaderLines(chunker, mimeType, 2)
+  let issues = validateGeneInHeader(headers)
+
+  const dataObj = {} // object to track multi-line validation concerns
+  await chunker.iterateLines((line, lineNum, isLastLine) => {
+    const parsedLine = parseLine(line, delimiter)
+    issues = issues.concat(validateValuesAreNumeric(parsedLine, isLastLine, lineNum, dataObj))
+    issues = issues.concat(validateColumnNumber(parsedLine, headers, lineNum, isLastLine, dataObj))
+    issues = issues.concat(validateUniqueCellNamesWithinFile(parsedLine, isLastLine, dataObj))
+    issues = issues.concat(validateMetadataLabelMatches(headers, parsedLine, isLastLine, dataObj))
+    issues = issues.concat(validateGroupColumnCounts(headers, parsedLine, isLastLine, dataObj))
+    // add other line-by-line validations here
```

Comment:
Similar to how we say `headers` instead of `parsedHeaders`, I'd suggest we refer to the lines we predominantly use in this high-level module as simply `lines` instead of `parsedLines`.  That's also consistent with how this module refers to the return value of `parseLine` elsewhere.

```suggestion
  await chunker.iterateLines((rawLine, lineNum, isLastLine) => {
    const line = parseLine(rawLine, delimiter)
    issues = issues.concat(validateValuesAreNumeric(line, isLastLine, lineNum, dataObj))
    issues = issues.concat(validateColumnNumber(line, headers, lineNum, isLastLine, dataObj))
    issues = issues.concat(validateUniqueCellNamesWithinFile(line, isLastLine, dataObj))
    issues = issues.concat(validateMetadataLabelMatches(headers, line, isLastLine, dataObj))
    issues = issues.concat(validateGroupColumnCounts(headers, line, isLastLine, dataObj))
    // add other line-by-line validations here
```